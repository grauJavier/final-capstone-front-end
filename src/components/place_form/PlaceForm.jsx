import { useForm } from 'react-hook-form';
import { useDispatch, useSelector } from 'react-redux';
import { useEffect } from 'react';
import ImageUploader from '../utils/UploadFile';
import fetchCities from '../../redux/reservation_form/actions/fetchCities';
import { sendPlaces } from '../../redux/places/placesSlice';

const PlaceForm = () => {
  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm();
  const dispatch = useDispatch();
  const cities = useSelector((state) => state.reservation.cities);
  const user = useSelector((state) => state.user.currentUser);

  useEffect(() => {
    dispatch(fetchCities());
  }, [dispatch]);

  const onSubmit = (data) => {
    dispatch(sendPlaces({ user_id: user.id, body: data }));
  };

  return (
    <div className="lg:ml-[15%] bg-turquoise-img bg-cover min-h-screen flex flex-col items-center justify-center p-4 pt-20">
      <div className="container flex flex-col items-center justify-center sm:p-24 sm:pb-40">
        <h1 className="secondary-font text-4xl text-zinc-50 mb-12 font-bold text-center">
          Add new place
        </h1>
        <form
          className="flex md:flex-row md:!items-start gap-16 forms-layout "
          onSubmit={handleSubmit(onSubmit)}
        >
          <div className="flex flex-col gap-5">
            <input
              {...register('place.name', {
                required: { value: true, message: 'Name is required' },
              })}
              placeholder="Name"
              type="text"
              className="w-64 md:w-80 forms-inputs bg-turquoise-blue"
            />
            {errors.place?.name?.message && (
              <p className="text-white italic">*{errors.place.name.message}</p>
            )}

            <select
              {...register('place.city_id', {
                required: { value: true, message: 'Select a location' },
              })}
              id="citySelect"
              placeholder="Select Location"
              className="w-64 md:w-80 forms-inputs bg-turquoise-blue"
            >
              <option value="">Select Location</option>
              {cities.map((city, index) => (
                <option key={index} value={city.id}>
                  {city.name}
                </option>
              ))}
            </select>
            {errors.place?.city_id?.message && (
              <p className="text-white italic">*{errors.place.city_id.message}</p>
            )}

            <textarea
              {...register('place.description', {
                required: { value: true, message: 'Description is required' },
              })}
              placeholder="Add a description"
              className="w-64 md:w-80 forms-inputs bg-turquoise-blue rounded-lg h-48"
              maxLength="255"
            />
            {errors.place?.description?.message && (
              <p className="text-white italic">*{errors.place.description.message}</p>
            )}

            <ImageUploader setValue={setValue} />
          </div>
          <div className="flex flex-col gap-5 w-64 md:w-80">
            <select
              {...register('details.place_type', {
                required: { value: true, message: 'Introduce a place type' },
              })}
              placeholder="Place type"
              className="w-full forms-inputs bg-turquoise-blue "
              id="citySelect"
            >
              <option value="">Place type</option>
              <option value="Room">Room</option>
              <option value="Entire Home">Entire Home</option>
            </select>
            {errors.details?.place_type?.message && (
              <p className="text-white italic">*{errors.details.place_type.message}</p>
            )}

            <select
              {...register('details.property_type', {
                required: { value: true, message: 'Introduce a property type' },
              })}
              placeholder="Property type"
              className="w-full forms-inputs bg-turquoise-blue"
              id="citySelect"
            >
              <option value="">Property type</option>
              <option value="House">House</option>
              <option value="Apartment">Apartment</option>
              <option value="Guest House">Guest House</option>
              <option value="Hotel">Hotel</option>
            </select>
            {errors.details?.property_type?.message && (
              <p className="text-white italic">*{errors.details.property_type.message}</p>
            )}

            <div className="flex flex-row items-center justify-between">
              <label htmlFor="number-of-bedrooms" className="text-white mr-2">
                Number of bedrooms
              </label>
              <input
                {...register('details.bedrooms', {
                  required: { value: true, message: 'Introduce the amount of bedrooms' },
                  validate: {
                    isInteger: (value) =>
                      Number.isInteger(Number(value)) || 'Please enter an integer',
                  },
                  min: { value: 1, message: 'Minimum value is 1' },
                })}
                id="number-of-bedrooms"
                type="number"
                min="1"
                defaultValue={1}
                className="forms-inputs bg-turquoise-blue w-20"
              />
              {errors.details?.bedrooms?.message && (
                <p className="text-white italic">*{errors.details.bedrooms.message}</p>
              )}
            </div>
            <div className="flex flex-row items-center justify-between">
              <label htmlFor="number-of-beds" className="text-white mr-2">
                Number of beds
              </label>
              <input
                {...register('details.beds', {
                  required: { value: true, message: 'Introduce the number of beds' },
                  validate: {
                    isInteger: (value) =>
                      Number.isInteger(Number(value)) || 'Please enter an integer',
                  },
                  min: { value: 1, message: 'Minimum value is 1' },
                })}
                id="number-of-beds"
                type="number"
                min="1"
                defaultValue={1}
                className="forms-inputs bg-turquoise-blue w-20"
              />
              {errors.details?.beds?.message && (
                <p className="text-white italic">*{errors.details.beds.message}</p>
              )}
            </div>
            <div className="flex flex-row items-center justify-between">
              <label htmlFor="number-of-bathrooms" className="text-white mr-2">
                Number of bathrooms
              </label>
              <input
                {...register('details.bathrooms', {
                  required: { value: true, message: 'Introduce the number of bathrooms' },
                  validate: {
                    isInteger: (value) =>
                      Number.isInteger(Number(value)) || 'Please enter an integer',
                  },
                  min: { value: 1, message: 'Minimum value is 1' },
                })}
                id="number-of-bathrooms"
                type="number"
                min="1"
                defaultValue={1}
                className="forms-inputs bg-turquoise-blue w-20"
              />
              {errors.details?.bathrooms?.message && (
                <p className="text-white italic">*{errors.details.bathrooms.message}</p>
              )}
            </div>
            <div className="flex flex-row items-center justify-between">
              <label htmlFor="price" className="text-white mr-5">
                Price
              </label>
              <input
                {...register('details.price', {
                  required: { value: true, message: 'Introduce the price' },
                  validate: {
                    isInteger: (value) =>
                      Number.isInteger(Number(value)) || 'Please enter an integer',
                  },
                  min: 1,
                })}
                id="price"
                type="number"
                defaultValue={0}
                min="0"
                className="w-full bg-turquoise-blue forms-inputs"
              />
              {errors.details?.price?.message && (
                <p className="text-white italic">*{errors.details.price.message}</p>
              )}
            </div>
            <input
              type="submit"
              className="forms-submit text-turquoise-blue hover:bg-turquoise-blue hover:text-white min-w-fit md:self-end self-center"
              value="Add new place"
            />
          </div>
        </form>
      </div>
    </div>
  );
};

export default PlaceForm;
