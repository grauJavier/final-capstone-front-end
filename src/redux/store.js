import { configureStore, combineReducers } from '@reduxjs/toolkit';
import userReducer from './user/userSlice';
import { persistStore, persistReducer } from 'redux-persist';
import sessionStorage from 'redux-persist/lib/storage/session'; // defaults to localStorage for web
import { createTransform } from 'redux-persist';
import placesSlice from "./places/placesSlice";
import reservationReducer from './reservation_form/reservationSlice.js';
import myReservationsReducer from './my_reservations/myReservationsSlice.js';
import noticeAlertReducer from './notice_alert/noticeAlertSlice.js';

// Create a personalized transform.
const userTransform = createTransform(
  (inboundState) => {
    return {
      currentUser: inboundState.currentUser,
      token: inboundState.token
    };
  },

  (outboundState) => {
    return {
      ...outboundState
    };
  },
  // Apply this transform only to the user state.
  { whitelist: ['user'] }
);

const persistConfig = {
  key: 'root',
  storage: sessionStorage,
  whitelist: ['user'],
  transforms: [userTransform]
};

const reducer = combineReducers({
  noticeAlert: noticeAlertReducer,
  user: userReducer,
  placesSlice,
  reservation: reservationReducer,
  myReservations: myReservationsReducer,
});

const persistedReducer = persistReducer(persistConfig, reducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false
    }),
});

const persistor = persistStore(store);

export { store, persistor };
