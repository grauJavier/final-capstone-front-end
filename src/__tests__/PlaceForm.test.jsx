import '@testing-library/jest-dom'; // For better DOM assertions
import { beforeEach, expect, describe, test } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes, Navigate } from 'react-router-dom';
import PlaceForm from '../components/place_form/PlaceForm';
import { store } from '../redux/store';

describe('place form', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Routes>
            <Route index path="/" element={<Navigate to="/new-place" />} />
            <Route path="/new-place" element={<PlaceForm />} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
  });

  test('should display the correct elements', async () => {
    const nameInput = screen.getByPlaceholderText('Name');
    const locationInput = screen.getByPlaceholderText('Select Location');
    const descriptionInput = screen.getByPlaceholderText('Add a description');
    const placeTypeInput = screen.getByPlaceholderText('Place type');
    const propertyTypeInput = screen.getByPlaceholderText('Property type');

    const button = screen.getByRole('button', { name: 'Add new place' });

    expect(nameInput).toBeInTheDocument();
    expect(locationInput).toBeInTheDocument();
    expect(descriptionInput).toBeInTheDocument();
    expect(placeTypeInput).toBeInTheDocument();
    expect(propertyTypeInput).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });

  test('should render correctly', async () => {
    await waitFor(() => expect(PlaceForm).toMatchSnapshot());
  });
});
