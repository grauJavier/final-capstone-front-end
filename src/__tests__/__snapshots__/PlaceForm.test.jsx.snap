// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`place form > should render correctly 1`] = `[Function]`;
